cassandra/src/main/scala/zio/flow/internal/CassandraKeyValueStore.scala

```diff
@@ -0,0 +1,172 @@
+package zio.flow.internal
+
+import com.datastax.oss.driver.api.core.cql.{AsyncResultSet, Row, SimpleStatement}
+import com.datastax.oss.driver.api.core.{CqlIdentifier, CqlSession}
+import com.datastax.oss.driver.api.querybuilder.QueryBuilder.literal
+import com.datastax.oss.driver.api.querybuilder.QueryBuilder
+
+import java.io.IOException
+import java.nio.ByteBuffer
+import zio.{Chunk, Has, IO, URLayer, ZIO}
+import zio.stream.ZStream
+
+final class CassandraKeyValueStore(session: CqlSession) extends KeyValueStore {
+  import CassandraKeyValueStore._
+
+  private val keyspace =
+    session.getKeyspace
+      .orElse(null: CqlIdentifier) // scalafix:ok DisableSyntax.null
+
+  private val table =
+    CqlIdentifier.fromCql(CassandraKeyValueStore.tableName)
+
+  private val cqlSelect =
+    QueryBuilder.selectFrom(keyspace, table)
+
+  private val cqlInsert =
+    QueryBuilder.insertInto(keyspace, table)
+
+  override def put(
+    namespace: String,
+    key: Chunk[Byte],
+    value: Chunk[Byte]
+  ): IO[IOException, Boolean] = {
+    val insert = cqlInsert
+      .value(
+        namespaceColumnName,
+        literal(namespace)
+      )
+      .value(
+        keyColumnName,
+        byteBufferFrom(key)
+      )
+      .value(
+        valueColumnName,
+        byteBufferFrom(value)
+      )
+      .build
+
+    executeAsync(insert, session)
+      .mapBoth(
+        refineToIOException(s"Error putting key-value pair for <$namespace> namespace"),
+        _ => true
+      )
+  }
+
+  override def get(namespace: String, key: Chunk[Byte]): IO[IOException, Option[Chunk[Byte]]] = {
+    val query = cqlSelect
+      .column(valueColumnName)
+      .whereColumn(namespaceColumnName)
+      .isEqualTo(
+        literal(namespace)
+      )
+      .whereColumn(keyColumnName)
+      .isEqualTo(
+        byteBufferFrom(key)
+      )
+      .limit(1)
+      .build
+
+    executeAsync(query, session).flatMap { result =>
+      if (result.remaining > 0)
+        ZIO.some(
+          blobValueOf(valueColumnName, result.one)
+        )
+      else
+        ZIO.none
+    }.mapError(
+      refineToIOException(s"Error retrieving or reading value for <$namespace> namespace")
+    )
+  }
+
+  override def scanAll(namespace: String): ZStream[Any, IOException, (Chunk[Byte], Chunk[Byte])] = {
+    val query = cqlSelect
+      .column(keyColumnName)
+      .column(valueColumnName)
+      .whereColumn(namespaceColumnName)
+      .isEqualTo(
+        literal(namespace)
+      )
+      .allowFiltering
```

Comment:
@adamgfraser There is a performance implication with this new data model:
```cql
CREATE TABLE _zflow_key_value_store (
  namespace   VARCHAR,
  key         BLOB,
  value       BLOB,
  PRIMARY KEY ((namespace, key))
)
```
Since `namespace` and `key` now form the compound partition key, retrieval now requires both columns:
```cql
SELECT value FROM _zflow_key_value_store WHERE namespace = 'ns_value' AND key = <bytes>;
```
This is essentially what `get` method does.

For `scanAll`, we are retrieving all values associated with a `namespace`. Since `namespace` is just one component of the compound key, Cassandra doesn't allow a query like `'SELECT key, value FROM _zflow_key_value_store WHERE namespace = 'ns_value''`, due to potential performance issue, unless `'ALLOW FILTERING'` is explicitly added to the query.

So for users of a Cassandra backend, they should be aware of such performance implication. However, I suppose the name `scanAll` itself might already imply some potential performance concern.

Reply:
I'm going to make the `key` column a clustering key, as in:
```cql
CREATE TABLE _zflow_key_value_store (
  namespace   VARCHAR,
  key         BLOB,
  value       BLOB,
  PRIMARY KEY (namespace, key)
)
```
Note the subtle difference on the Primary Key definition. Now the partition key is just the `namespace` column and the clustering key is the `key` column. This will solve the performance issue mentioned earlier.

Reply:
Updated as mentioned above to avoid "allow filtering."